File: book/snippets/nostr/python/src/nip21.py
from nostr_protocol import Keys, PublicKey, EventBuilder, EventId, Nip21, Nip19Profile, Nip19Event, Kind, Coordinate


def nip21():
    print()
    print("Nostr URIs:")
    # ANCHOR: npub
    keys = Keys.generate()

    # URI npub
    pk_uri = keys.public_key().to_nostr_uri()
    print(f" Public key (URI):    {pk_uri}")

    # bech32 npub
    pk_parse = Nip21.parse(pk_uri)
    if pk_parse.as_enum().is_pubkey():
        pk_bech32 = PublicKey.from_nostr_uri(pk_uri).to_bech32()
        print(f" Public key (bech32): {pk_bech32}")
    # ANCHOR_END: npub

    print()

    # ANCHOR: note
    event = EventBuilder.text_note("Hello from Rust Nostr Python bindings!", []).to_event(keys)

    # URI note
    note_uri = event.id().to_nostr_uri()
    print(f" Event (URI):    {note_uri}")

    # bech32 note
    note_pasre = Nip21.parse(note_uri)
    if note_pasre.as_enum().is_note():
        event_bech32 = EventId.from_nostr_uri(note_uri).to_bech32()
        print(f" Event (bech32): {event_bech32}")
    # ANCHOR_END: note

    print()

    # ANCHOR: nprofile
    relays = ["wss://relay.damus.io"]
    nprofile = Nip19Profile(keys.public_key(), relays)

    # URI nprofile
    nprofile_uri = nprofile.to_nostr_uri()
    print(f" Profile (URI):    {nprofile_uri}")

    # bech32 nprofile
    nprofile_parse = Nip21.parse(nprofile_uri)
    if nprofile_parse.as_enum().is_profile():
        nprofile_bech32 = Nip19Profile.from_nostr_uri(nprofile_uri).to_bech32()
        print(f" Profile (bech32): {nprofile_bech32}")
    # ANCHOR_END: nprofile

    print()

    # ANCHOR: nevent
    relays = ["wss://relay.damus.io"]
    nevent = Nip19Event(event.id(), keys.public_key(), kind=None, relays=relays)

    # URI nevent
    nevent_uri = nevent.to_nostr_uri()
    print(f" Event (URI):    {nevent_uri}")

    # bech32 nevent
    nevent_parse = Nip21.parse(nevent_uri)
    if nevent_parse.as_enum().is_event():
        nevent_bech32 = Nip19Event.from_nostr_uri(nevent_uri).to_bech32()
        print(f" Event (bech32): {nevent_bech32}")
    # ANCHOR_END: nevent

    print()

    # ANCHOR: naddr
    coord = Coordinate(Kind(0), keys.public_key())

    # URI naddr
    coord_uri = coord.to_nostr_uri()
    print(f" Coordinate (URI):    {coord_uri}")

    # bech32 naddr
    coord_parse = Nip21.parse(coord_uri)
    if coord_parse.as_enum().is_coord():
        coord_bech32 = Coordinate.from_nostr_uri(coord_uri).to_bech32()
        print(f" Coordinate (bech32): {coord_bech32}")
    # ANCHOR_END: naddr

File: book/snippets/nostr/js/src/nip21.js
const { loadWasmSync, Keys, EventBuilder, Nip19Profile, Nip19Event, Coordinate} = require("@rust-nostr/nostr");

function run(){
    // Load WASM
    loadWasmSync();

    console.log();
    console.log("Nostr URIs:");

    // ANCHOR: npub
    let keys = Keys.generate();
    // bech32 npub
    let pk_bech32 = keys.publicKey.toBech32();
    console.log(` Public key (bech32): ${pk_bech32}`);
    // ANCHOR_END: npub

    console.log();
    // ANCHOR: note
    let event = EventBuilder.textNote("Hello from Rust Nostr JS Bindings!", []).toEvent(keys);

    // bech32 note
    let note_bech32 = event.id.toBech32()
    console.log(` Event (bech32): ${note_bech32}`);
    // ANCHOR_END: note

    console.log();
    // ANCHOR: nprofile
    let relays = ["wss://relay.damus.io"];
    let nprofile = new Nip19Profile(keys.publicKey, relays);

    // URI nprofile
    let nprofile_uri = nprofile.toNostrUri();
    console.log(` Profile (URI):    ${nprofile_uri}`);

    // bech32 nprofile
    let nprofile_bech32 = Nip19Profile.fromNostrUri(nprofile_uri).toBech32();
    console.log(` Profile (bech32): ${nprofile_bech32}`);
    // ANCHOR_END: nprofile
    
    console.log();
    // ANCHOR: nevent
    let nevent = new Nip19Event(event.id, keys.publicKey, null, relays);

    // URI nevent
    let nevent_uri = nevent.toNostrUri();
    console.log(` Event (URI):    ${nevent_uri}`);

    // bech32 nevent
    let nevent_bech32 = Nip19Event.fromNostrUri(nevent_uri).toBech32();
    console.log(` Event (bech32): ${nevent_bech32}`);
    // ANCHOR_END: nevent

    console.log();
    // ANCHOR: naddr
    // URI naddr
    let coord_uri = new Coordinate(event.kind, keys.publicKey).toNostrUri();
    console.log(` Coordinate (URI):    ${coord_uri}`);

    // bech32 naddr
    let coord_bech32 = new Coordinate(event.kind, keys.publicKey).toBech32();
    console.log(` Coordinate (bech32): ${coord_bech32}`);
    // ANCHOR_END: naddr

}

run();